src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -268,10 +274,35 @@ async def setup_user_lockfile_requests(
     )
 
 
+class PythonSyntheticLockfilesTargetAdaptorsRequest(SyntheticTargetAdaptorsRequest):
+    pass
+
+
+@rule
+async def python_lockfile_synthetic_targets(
+    _: PythonSyntheticLockfilesTargetAdaptorsRequest,
+    python_setup: PythonSetup,
+) -> SyntheticTargetAdaptors:
+    return SyntheticTargetAdaptors(
+        [
+            SyntheticTargetAdaptor.create(
+                path=os.path.dirname(lockfile),
+                type_alias="lockfiles",
+                name=name,
+                sources=[os.path.basename(lockfile)],
+            )
+            for name, lockfile in python_setup.resolves.items()
+        ]
+        if python_setup.enable_resolves
+        else []
+    )
```

Comment:
This is an interesting example I think, because I would expect the request to contain the file to generate a target for... via maybe a sources glob that drives which files are fed to each implementation?

I suppose that you wanted to allow for synthetic targets which were not associated with any file? It would be interesting to see the potential use cases for that, because we've definitely been trending further toward "everything is a file", and biasing toward file-centric APIs might make sense.

If this API were going to be used to generate _all_ python targets in the repository (for example), an API closer to tailor's, where the request specifies which files to generate targets for might make more sense? That would push more of the concern for scanning the repository on the caller of the plugin, rather than on the plugin itself.

Reply:
My goal with the synthetic targets where to allow programmatically create targets, as if they were declared in a BUILD file.

I think I agree with your point, but that feels like a longer goal, and this could perhaps be a stop on the way there?
In my attempts to create a synthesized file target I failed, due to #17092 

In this example, there is no source file, as we're just going to assume it exists, as it should in case generate-lockfiles is run successfully. Perhaps better would be to first fetch the file in question with a source glob and then create the target only if it actually exists. OTOH if the lockfile doesn't exist that's an issue too, and this target would help surface that issue.

Reply:
> I think I agree with your point, but that feels like a longer goal, and this could perhaps be a stop on the way there?
In my attempts to create a synthesized file target I failed, due to https://github.com/pantsbuild/pants/pull/17092

Yea... I think that you're right. You're hooking the relevant bits of the code... there just isn't really a filter happening on paths? I'll review that one.

Thanks!